README.md
# README

[ldu-devops-docker](https://lildutils.hu/projects/ldu-devops-docker/about.html)

---

## Overview

Overview is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

---

## Development

Development is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

### PreRequisites

- [Git](https://git-scr.com/download) - git version ^2.34.1
- [GNU Make](https://www.gnu.org/software/make/#download) - GNU Make ^4.3

### Watcher

Watcher Task is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. 

``` sh
make watch
```

---

## Test

Test is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

### Unit

Unit Test Task is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. 

``` sh
make test
```

### End-to-End

End to End Test Task is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. 

``` sh
make e2e
```

---

## Build

Build is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

Build Task is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. 

``` sh
make build
```

---

## Getting Started

Getting Started is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

### Services

Services is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 

### *List of Docker Compose Services*

|     | **Service**  | **Image**                                | **Port(s)** |
|----:|:-------------|:-----------------------------------------|------------:|
|   1 | nginx        | jc21/nginx-proxy-manager:2.10.0          | 80,443,6061 |
|   2 | portainer    | portainer/portainer-ce:2.17.1-alpine     |        6062 |
|   3 | vscode       | linuxserver/code-server:4.11.0           |        6063 |
|   4 | mysql        | mysql:8-debian                           |  3306,33060 |
|   5 | phpmyadmin   | phpmyadmin:5.2.1-apache                  |        6064 |
|   6 | postgres     | postgres:15-alpine                       |        5432 |
|   7 | pgadmin      | dpage/pgadmin4:6.21                      |        6065 |
|   8 | mongodb      | mongo:4-focal                            |       27017 |
|   9 | mongoxpress  | mongo-express:1.0.0-alpha                |        6066 |
|  10 | gitlab       | gitlab/gitlab-ce:15.10.0-ce.0            |          80 |
|  11 | gitlabrunner | gitlab/gitlab-runner:alpine3.15-v15.10.0 |           - |
|  12 | sonarqubedb  | postgres:15-alpine                       |        5432 |
|  13 | sonarqube    | sonarqube:9.9.0-community                |        9000 |

### Auto-Install

Auto-Install is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 

``` sh
bash <(curl -sL https://raw.githubusercontent.com/lildutils/ldu-devops-docker/master/dist/latest/make.sh)
```

### Download to AnyWhere

Download to AnyWhere is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 

``` sh
curl -sL https://raw.githubusercontent.com/lildutils/ldu-devops-docker/master/dist/latest/make.sh -o /path/to/make.sh
```

### Configurations

Configurations is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 

#### Application

Application Configurations is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.

``` ini
CRON_BACKUP_INSTANCES=comma separated list of: devops,portainer,nginx,vscode,mysql,postgres,mongodb,gitlab,sonarqube
CRON_BACKUP_MODES=comma separated list of: daily,weekly,monthly
DEBUG_MODE=showing debug informations: TRUE|FALSE
DIR_DEVOPS=devops directory path
DIR_DEVOPS_BACKUPS=backups directory path
DIR_DEVOPS_DATA=data directory path
FAIL2BAN_MAX_RETRY=integer value to maximalize the login retries
FILE_DEVOPS_ENV=url to auto-download docker-compose .env file
FILE_DEVOPS_COMPOSE=url to auto-download docker-compose .yml file
FILE_DEVOPS_MAKE=url to auto-download makefile
UFW_ALLOW_IN=comma separated list of ports (f.e.: 80,443,22)
UFW_ALLOW_OUT=comma separated list of ports (f.e.: 80,443,53,631)
```

#### Docker Compose Services

Docker Compose Service Configurations is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.

``` env
DOMAIN=domain name of the server
GITLAB_ROOT_USER=username
GITLAB_ROOT_PASSWORD=password
GITLAB_EXTERNAL_URL=http(s) schema url of the gitlab service
GITLAB_EXTERNAL_DOMAIN=only domain name of the gitlab service
GITLAB_RUNNER_REGISTRATION_TOKEN=runner registration token of gitlab service
GITLAB_SMTP_ENABLED=true|false, if true the smpt server is enabled in gitlab service
GITLAB_SMTP_SERVER=smtp server domain
GITLAB_SMTP_PORT=465|587|25
GITLAB_SMTP_USER=useremail
GITLAB_SMTP_PASSWORD=password
GITLAB_SMTP_DOMAIN=smtp server allowed domain
GITLAB_SMTP_FROM=gitlab@smtp server allowed domain
GITLAB_SMTP_REPLY=noreply@smtp server allowed domain
MONGO_ROOT_USER=username
MONGO_ROOT_PASSWORD=password
MYSQL_ROOT_USER=username
MYSQL_ROOT_PASSWORD=password
NGINX_ROOT_USER=admin@example.com
NGINX_ROOT_PASSWORD=changeme
PORTAINER_ROOT_USER=username
PORTAINER_ROOT_PASSWORD=password
POSTGRES_ROOT_USER=username
POSTGRES_ROOT_PASSWORD=password
POSTGRES_PGADMIN_USER=useremail
POSTGRES_PGADMIN_PASSWORD=password
SONARQUBE_ROOT_USER=admin
SONARQUBE_ROOT_PASSWORD=admin
SONARQUBE_DB_USER=sonar
SONARQUBE_DB_PASSWORD=password
VSCODE_ROOT_USER=username
VSCODE_ROOT_PASSWORD=password
VSCODE_SUDO_USER=username
VSCODE_SUDO_PASSWORD=password
VSCODE_SUDO_PUID=user id
VSCODE_SUDO_PGID=group id
VSCODE_WORKSPACE=path of the user home directory
```

### Usage

Usage is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 

``` sh
/path/to/make.sh --help
```

### LDT compatibility

LDT compatibility is Coming soon... Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 

``` sh
ldt devops --help
```

---
###### Have a good DevOPSing! (:

##### Bests,
##### d


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.


## [1.0.0](https://github.com/lildutils/ldu-devops-docker/releases/tag/1.0.0) (2023-04-xx)

- FEAT: Nginxdb service in docker compose
- FIX: Docker compose volume refs
- FIX: Docker Compose file for full devops
- FEAT: Install vscode container post script created
- FEAT: Docker compose services part #1
- CHORE: Manual deploy
- FIX: Make file copy to devops
- FIX: Docker compose environment urls
- FIX: Bash application script auto download urls
- FIX: Bash application script
- FIX: Bash compiler lib
- FEAT: Test Framework next steps
- FEAT: Application compiler for make.sh script file
- FEAT: Nginx docker compose service
- FEAT: Autorun for make.sh
- FEAT: Wait for check logs and than clear terminal
- FEAT: Dynamic database service support
- FEAT: Edit docker environment before compose up
- FEAT: dynamic ufw allowed rules
- FEAT: Clean terminal after command run
- FEAT: Fail2Ban max retry dynamic setting
- FEAT: Gitlab SMTP enabled or not
- FEAT: Build process with clean option
- FEAT: README markdown file created
- FEAT: MIT Licensing
- FEAT: Ignoring files from git
- FEAT: Initial testing frame
- FEAT: Docker resources for make devops
- FEAT: Make devops script created

## [0.0.0](https://github.com/lildutils/ldu-devops-docker/releases/tag/0.0.0) (2023-03-28)

- FEAT: Project file created
- FEAT: Initial commit


src/main/sh/make.sh
#!/bin/bash

main() {
    COMMAND=$1
    if [ "$1" == "" ]; then COMMAND=install; fi
    CURRDIR=$(realpath $(dirname "${BASH_SOURCE[0]}")/)
    _loadEnvironment
    shift
    $COMMAND $*
    exit 0
}

_loadEnvironment() {
    ## load default environment
    DIR_DEVOPS=/devops
    DIR_DEVOPS_BACKUPS=/backups
    DIR_DEVOPS_DATA=/data
    DOCKER_INIT_DB_CLIENTS=phpmyadmin,pgadmin,mongoxpress
    FAIL2BAN_MAX_RETRY=3
    FILE_DEVOPS_ENV=https://raw.githubusercontent.com/lildutils/ldu-devops/master/dist/latest/resources/docker-compose.env.example
    FILE_DEVOPS_COMPOSE=https://raw.githubusercontent.com/lildutils/ldu-devops/master/dist/latest/resources/docker-compose.yml
    GITLAB_RUNNER_REGISTRATION_TOKEN=registration-token
    UFW_ALLOW_IN=http,https,ssh
    UFW_ALLOW_OUT=http,https,53/udp

    ## override environment if .ini file exists
    INIFILE=$(realpath $(dirname "${BASH_SOURCE[0]}")/make.ini)
    if [ -f $INIFILE ]; then
        export $(cat $INIFILE | xargs)
    fi
}

install() {
    _doInstallTools
    _doInstallDocker
    _doInstallDockerCompose
    _doInstallDevOps
    _doDockerComposeUp
    _doDockerGCC
    _doTerminalClean
}

backup() {
    currentBackups=$DIR_DEVOPS_BACKUPS/$(echo $(date +"%Y%m%d%H%M%S"))

    devopsBackups=$currentBackups/devops/docker
    nginxBackups=$currentBackups/nginx/configs
    portainerBackups=$currentBackups/portainer/exports
    dbBackups=$currentBackups/db/dumps

    mkdir -p $devopsBackups/
    mkdir -p $nginxBackups/
    mkdir -p $portainerBackups/
    mkdir -p $dbBackups/

    ## devops
    cp -r $DIR_DEVOPS/.env $devopsBackups/docker-compose.env.backup
    cp -r $DIR_DEVOPS/docker-compose.yml $devopsBackups/docker-compose.yml.backup

    ## TODO: nginx

    ## TODO: portainer

    ## TODO: databases

    _doTerminalClean
}

clean() {
    _doDockerClean

    rm -rf $DEVOPS_DATA_DIR/
    rm -rf $DEVOPS_ROOT_DIR/

    _doTerminalClean
}

registerGitlabRunner() {
    docker exec -it gitlabrunner \
        gitlab-runner register --non-interactive \
        --registration-token $GITLAB_RUNNER_REGISTRATION_TOKEN \
        --name docker-gitlab-runner \
        --url http://gitlab \
        --tag-list docker,docker-stable \
        --executor docker \
        --docker-image docker:stable

    _doTerminalClean
}

_doInstallTools() {
    apt-get update -y
    apt-get full-upgrade -y
    apt-get autoremove -y

    apt-get install -y curl htop mc nano net-tools ufw fail2ban

    systemctl start ufw
    systemctl enable ufw
    ufw enable
    ufw default deny incoming
    ufw default deny outgoing
    for rule in ${UFW_ALLOW_IN//,/ }; do
        ufw allow in $rule
    done
    for rule in ${UFW_ALLOW_OUT//,/ }; do
        ufw allow out $rule
    done
    ufw reload

    systemctl start fail2ban
    systemctl enable fail2ban
    echo "[sshd]" >/etc/fail2ban/jail.local
    echo "enabled = true" >>/etc/fail2ban/jail.local
    echo "port = 22" >>/etc/fail2ban/jail.local
    echo "filter = sshd" >>/etc/fail2ban/jail.local
    echo "logpath = /var/log/auth.log" >>/etc/fail2ban/jail.local
    echo "maxretry = $FAIL2BAN_MAX_RETRY" >>/etc/fail2ban/jail.local
    systemctl restart fail2ban
}

_doInstallDocker() {
    mkdir -p ./tmp/
    curl -sL "https://get.docker.com" -o ./tmp/install-docker.sh
    chmod 755 ./tmp/install-docker.sh
    /bin/sh ./tmp/install-docker.sh
    rm -rf ./tmp/install-docker.sh
}

_doInstallDockerCompose() {
    apt-get install -y docker-compose
}

_doInstallDevOps() {
    if [ ! -d "$DIR_DEVOPS/" ]; then mkdir -p $DIR_DEVOPS/; fi
    if [ ! -d "$DIR_DEVOPS_BACKUPS/" ]; then mkdir -p $DIR_DEVOPS_BACKUPS/; fi
    if [ ! -d "$DIR_DEVOPS_DATA/" ]; then mkdir -p $DIR_DEVOPS_DATA/; fi

    rm -rf $DIR_DEVOPS/docker-compose.yml

    if [ ! -f $DIR_DEVOPS/.env ]; then
        curl -sL "$FILE_DEVOPS_ENV" -o $DIR_DEVOPS/.env
        nano $DIR_DEVOPS/.env
    fi
    curl -sL "$FILE_DEVOPS_COMPOSE" -o $DIR_DEVOPS/docker-compose.yml
}

_doDockerComposeUp() {
    _doDockerComposeUpPre

    docker-compose \
        --project-name devops \
        --project-directory $DIR_DEVOPS \
        --file $DIR_DEVOPS/docker-compose.yml \
        up -d \
        --remove-orphans

    _doDockerComposeUpPost
}

_doDockerComposeUpPre() {
    if [ -f $DIR_DEVOPS/.env ]; then export $(cat $DIR_DEVOPS/.env | xargs); fi

    for dbClient in ${DOCKER_INIT_DB_CLIENTS//,/ }; do
        docker-compose --project-name devops --project-directory $DIR_DEVOPS --file $DIR_DEVOPS/docker-compose.yml up -d $dbClient

        if [ "$dbClient" == "phpmyadmin" ]; then
            mkdir -p ./tmp/

            echo "CREATE USER '$PROXYMNGR_DB_USER'@'%' IDENTIFIED BY '$PROXYMNGR_DB_PASSWORD';" >./tmp/mysql-init.sql
            echo "CREATE DATABASE proxymanager;" >>./tmp/mysql-init.sql
            echo "GRANT ALL PRIVILEGES ON proxymanager.* TO '$PROXYMNGR_DB_USER'@'%';" >>./tmp/mysql-init.sql

            sleep 30 # TODO: wait for service 'mysql' has started, avoid sleep
            docker exec -i mysql sh -c "exec mysql -u$MYSQL_ROOT_USER -p'$MYSQL_ROOT_PASSWORD'" <./tmp/mysql-init.sql

            rm -rf ./tmp/mysql-init.sql
        fi

        if [ "$dbClient" == "pgadmin" ]; then
            mkdir -p ./tmp/

            echo "CREATE ROLE $SONARQUBE_SONAR_USER WITH LOGIN NOSUPERUSER INHERIT NOCREATEDB NOCREATEROLE NOREPLICATION PASSWORD '$SONARQUBE_SONAR_PASSWORD';" >./tmp/postgres-init.sql
            echo "CREATE DATABASE sonar WITH OWNER = sonar;" >>./tmp/postgres-init.sql
            echo "GRANT ALL PRIVILEGES ON DATABASE sonar TO $SONARQUBE_SONAR_USER;" >>./tmp/postgres-init.sql

            sleep 30 # TODO: wait for service 'postgres' has started, avoid sleep
            docker exec -i postgres sh -c "exec psql -u$POSTGRES_ROOT_USER -p$POSTGRES_ROOT_PASSWORD" <./tmp/postgres-init.sql

            rm -rf ./tmp/postgres-init.sql
        fi
    done
}

_doDockerComposeUpPost() {
    docker --version
    docker-compose --version
}

_doDockerGCC() {
    echo y | docker system prune
}

_doDockerClean() {
    docker container kill $(docker container ls -aq)
    docker container rm --force $(docker container ls -aq)
    docker volume rm --force $(docker volume ls -aq)
    docker image rm --force $(docker image ls -aq)
    docker network rm --force $(docker network ls -aq)
}

_doTerminalClean() {
    read -r -p "Press any key to continue..." -t 5 -n 1 -s
    rm -rf ~/.cache/ ~/.config/ ~/.docker/ ~/.gnupg/ ~/.local/ ~/.nano/ ~/.bash_history ~/.selected_editor ~/.sudo_as_admin_successful
    history -c
    clear
}

main $*
